### PR TITLE
Shared server class

### PR DESCRIPTION
**Pull request by ItsDrike**
As [discussed on discord](https://discord.com/channels/936788458939224094/938623566964981802/942557145780588585), this adds a shared `MCServer` base class between both `JavaServer` and `BedrockServer`. This base class holds shared common functionalities between both classes, which were previously repeated in both classes or separated as standalone global functions.

Stalled by:
- [x] #222

### CODE REVIEW

File: mcstatus/server.py [hunk 1]
```diff
@@ -48,22 +52,48 @@ def port(self) -> int:
...
     @classmethod
     def lookup(cls, address: str, timeout: float = 3) -> Self:
-        """Parses the given address and checks DNS records for an SRV record that points to the Minecraft server.
+        """Mimics minecraft's server address field.
+
+        :param str address: The address of the Minecraft server, like `example.com:19132`
+        :param float timeout: The timeout in seconds before failing to connect.
+        """
+        addr = Address.parse_address(address, default_port=None)
+        return cls(addr.host, addr.port, timeout=timeout)
```
Comment (kevinkjt2000) [inline mcstatus/server.py:61]: Is it possible to reword this here and inherit the docstring in child classes? There's a typo in mimics (mimcs) down below.
```suggestion
        """Mimics behavior of Minecraft's server address field when adding a multiplayer server.

        :param str address: The address of the Minecraft server, like `example.com:19132`
        :param float timeout: The timeout in seconds before failing to connect.
        """
        raise NotImplementedError
```

File: mcstatus/server.py [hunk 2]
```diff
@@ -24,15 +25,18 @@
...
+    """Base abstract class for a general minecraft server.
 
-    :param str host: The host/address/ip of the Minecraft server.
+    This class only contains the basic logic shared across both java and bedrock versions,
+    it doesn't include any version specific settings and it can't be used to make any requests.
+
+    :param str host: The host/ip of the minecraft server.
     :param int port: The port that the server is on.
-    :param float timeout: The timeout in seconds before failing to connect.
+    :param float timeout: Timeout, in seconds, before failing to connect.
```
Comment (kevinkjt2000) [inline mcstatus/server.py:36]: Correcting the punctuation here, but not in other places that have the same sentence?

**Review by PerchunPak — APPROVED** [review]

**Comment by PerchunPak** [comment]
How about add `is_java()` or `is_bedrock()` method?

**Comment by ItsDrike** [comment]
> How about add `is_java()` or `is_bedrock()` method?

I'm not aware of any straight forward way to identify a Java server from a Bedrock one, other than actually pinging it. The shared server class is only supposed to be an abstraction for shared elements in both classes, and to make something things like isinstance checks easier.

This means the base class doesn't contain the specific logic on how to perform ping for java nor for bedrock, and it can't be included without introducing something like `java_ping` and `bedrock_ping` methods, which will not be done, because that would mean a `JavaServer` inheriting from this base class would suddenly have `bedrock_ping` method, which makes no sense.

These could potentially be implemented as external global methods, making use of the JavaServer and BedrockServer classes and simply performing a ping to identify the server type, but again, unless there's something I'm not aware of, I don't think this can be done without performing the full ping, so it's questionable whether this should be implemented at all.

Either way, it probably doesn't belong to this PR, if you want to see this implemented, you can open a standalone issue about it.

**Comment by PerchunPak** [comment]
I mean that if `MCServer` will have methods for cross-platform ping (as I wrote [in discord](https://discord.com/channels/936788458939224094/938591600160956446/971715699393302538)), this will help determine is server java or bedrock.
So I see it like
```py
class MCServer(ABC):
...
    @abstractmethod
    def is_java(self) -> bool:
        """Return bool is class instance of JavaServer."""
...
class JavaServer(MCServer):
...
    def is_java(self) -> bool: return True
...
class BedrockServer(MCServer):
...
    def is_java(self) -> bool: return False
...
```
But yes, it should be separate PR, and better to be created after this will be merged.

**Comment by kevinkjt2000** [comment]
`isinstance(object, JavaServer)` is something a user could do to save us the hassle of causing every class to update each time a new derivation of MCServer is implemented.

**Review by kevinkjt2000 — APPROVED** [review]

**Pull request merged**